libs/api/ebpf_api.cpp

```diff
@@ -1669,6 +1674,225 @@ _get_next_map_to_create(std::vector<ebpf_map_t*>& maps)
+    } else {
+        // String is in programs section (.dll files do this).
+        uintptr_t base = pe_context->image_base + section_header.VirtualAddress;
+        ebpf_assert(address >= base && address < base + section_header.Misc.VirtualSize);
+        uintptr_t offset = address - base;
+        return (const char*)(buffer->buf + offset);
+    }
+}
+
+static int
+_ebpf_pe_get_section_names(
+    void* context,
+    const VA& va,
+    const std::string& section_name,
+    const image_section_header& section_header,
+    const bounded_buffer* buffer)
+{
+    UNREFERENCED_PARAMETER(va);
+
+    ebpf_pe_context_t* pe_context = (ebpf_pe_context_t*)context;
+    if (section_name == "programs") {
+        // bpf2c generates a section that has ELF section names as strings at the
+        // start of the section.  Skip over them looking for the program_entry_t
+        // which starts with a 16-byte-aligned NULL pointer.
+        uint32_t program_offset = 0;
+        uint64_t zero = 0;
+        while (program_offset < section_header.Misc.VirtualSize &&
+               memcmp(buffer->buf + program_offset, &zero, sizeof(zero)) != 0) {
+            program_offset += 16;
+        }
+        int program_count = (section_header.Misc.VirtualSize - program_offset) / sizeof(program_entry_t);
+        for (int i = 0; i < program_count; i++) {
+            program_entry_t* program = (program_entry_t*)(buffer->buf + program_offset + i * sizeof(program_entry_t));
+            const char* pe_section_name =
+                _ebpf_get_section_string(pe_context, (uintptr_t)program->pe_section_name, section_header, buffer);
+            const char* elf_section_name =
+                _ebpf_get_section_string(pe_context, (uintptr_t)program->section_name, section_header, buffer);
+            pe_context->section_names[pe_section_name] = elf_section_name;
+        }
+    }
+
+    return 0;
+}
+
+static int
+_ebpf_pe_add_section(
+    void* context,
+    const VA& va,
+    const std::string& pe_section_name,
+    const image_section_header& section_header,
+    const bounded_buffer* buffer)
+{
+    UNREFERENCED_PARAMETER(va);
+    UNREFERENCED_PARAMETER(buffer);
+
+    if (!(section_header.Characteristics & IMAGE_SCN_CNT_CODE)) {
+        // Not a code section.
+        return 0;
+    }
+    ebpf_pe_context_t* pe_context = (ebpf_pe_context_t*)context;
+
+    // Get ELF section name.
+    if (!pe_context->section_names.contains(pe_section_name)) {
+        // Not an eBPF program section.
+        return 0;
+    }
+    std::string elf_section_name = pe_context->section_names[pe_section_name];
+
+    ebpf_section_info_t* info = (ebpf_section_info_t*)malloc(sizeof(*info));
+    if (info == nullptr) {
+        return 1;
+    }
+    memset(info, 0, sizeof(*info));
+    info->section_name = _strdup(elf_section_name.c_str());
+
+    info->program_type_name = _strdup(get_program_type_windows(elf_section_name, std::string()).name.c_str());
```

Comment:
`program_entry_t` in the PE file also contains the program type GUID (and attach type GUID). We can use that instead of recalculating the type.

Reply:
Done